File: python/47.py
import unittest
from typing import List, Set


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        return self.dfs(nums, set())

    def dfs(self, nums: List[int], used: Set[int]) -> List[List[int]]:
        ret = []

        for i in range(len(nums)):
            if i in used or (i > 0 and nums[i] == nums[i - 1] and i - 1 not in used):
                continue

            used.add(i)
            for next in self.dfs(nums, used):
                ret.append([nums[i]] + next)
            used.remove(i)

        if not ret:
            ret.append([])

        return ret


class TestSolution(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertListEqual(solution.permuteUnique([1, 1, 2]), [
                             [1, 1, 2], [1, 2, 1], [2, 1, 1]])

File: typescript/47.ts
function permuteUnique(nums: number[]): number[][] {
    nums.sort((a, b) => a - b);
    return dfs(nums, new Set());
};

const dfs = (nums: number[], used: Set<number>): number[][] => {
    const ret: number[][] = [];

    for (let i = 0; i < nums.length; i++) {
        if (used.has(i) || (i > 0 && nums[i] === nums[i - 1] && !used.has(i - 1))) continue;

        used.add(i);
        for (let next of dfs(nums, used)) ret.push([nums[i], ...next]);
        used.delete(i);
    }

    if (!ret.length) ret.push([]);

    return ret;
};

test('47', () => {
    expect(permuteUnique([1, 1, 2])).toEqual([
        [1, 1, 2],
        [1, 2, 1],
        [2, 1, 1],
    ]);
    expect(permuteUnique([1, 2, 3])).toEqual([
        [1, 2, 3],
        [1, 3, 2],
        [2, 1, 3],
        [2, 3, 1],
        [3, 1, 2],
        [3, 2, 1],
    ]);
});